config.txt
# quiet mode disables all visual output (for experiments).
quietMode = 0

# debug mode enables additional drawing and visualization.
debugMode = 1

# base path for video sequences.
sequenceBasePath = /Users/xiaofeidu/mData/visualTrackingBenchmark_v1.0

# path for output results file.
# comment this out to disable output.
resultsPath = /Users/xiaofeidu/workspace/struck/sequences/log.txt

# path for output precision file.
# comment this out to disable output
precisionPath = /Users/xiaofeidu/mData/visualTrackingBenchmark_v1.0

# video sequence to run the tracker on.
# comment this out to use webcam.
sequenceName = Basketball

# multi sequence names
#multiSeq = Biker
#multiSeq = BlurFace
#multiSeq = BlurOwl
#multiSeq = Box
#multiSeq = Car4
#multiSeq = CarScale
#multiSeq = Crowds
#multiSeq = ClifBar
#multiSeq = Couple
#multiSeq = Crossing
#multiSeq = David3
#multiSeq = Deer
#multiSeq = Diving
#multiSeq = Dog1
#multiSeq = Doll

multiSeq = Soccer
multiSeq = Subway
multiSeq = Surfer
multiSeq = Suv
multiSeq = Sylvester
#multiSeq = Tiger1
#multiSeq = Tiger2
multiSeq = Toy
multiSeq = Trellis
multiSeq = Twinnings
#multiSeq = Vase
multiSeq = Walking
multiSeq = Walking2
multiSeq = Woman

# frame size for use during tracking.
# the input image will be scaled to this size.
frameWidth = 320
frameHeight = 240

# seed for random number generator.
seed = 0

# tracker search radius in pixels.
searchRadius = 30

# SVM regularization parameter.
svmC = 100.0
# SVM budget size (0 = no budget).
svmBudgetSize = 50

# image features to use.
# format is: feature kernel [kernel-params]
# where:
#   feature = haar/raw/histogram
#   kernel = gaussian/linear/intersection/chi2
#   for kernel=gaussian, kernel-params is sigma
# multiple features can be specified and will be combined

#feature = haar gaussian 0.2
#feature = raw gaussian 0.1
#feature = histogram intersection
feature = circle_hsv intersection
#feature = hsv intersection  
#feature = circle_grad intersection   
#feature = circle_rgb intersection
#feature = circle_patch intersection

src/Config.cpp
/* 
 * Struck: Structured Output Tracking with Kernels
 * 
 * Code to accompany the paper:
 *   Struck: Structured Output Tracking with Kernels
 *   Sam Hare, Amir Saffari, Philip H. S. Torr
 *   International Conference on Computer Vision (ICCV), 2011
 * 
 * Copyright (C) 2011 Sam Hare, Oxford Brookes University, Oxford, UK
 * 
 * This file is part of Struck.
 * 
 * Struck is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Struck is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Struck.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */

#include "Config.h"

#include <fstream>
#include <iostream>
#include <sstream>

using namespace std;

Config::Config(const std::string& path)
{
	SetDefaults();
	
	ifstream f(path.c_str());
	if (!f)
	{
		cout << "error: could not load config file: " << path << endl;
		return;
	}
	
	string line, name, tmp;
	while (getline(f, line))
	{
		istringstream iss(line);
		iss >> name >> tmp;
		
		// skip invalid lines and comments
		if (iss.fail() || tmp != "=" || name[0] == '#') continue;
		
		if      (name == "seed") iss >> seed;
		else if (name == "quietMode") iss >> quietMode;
		else if (name == "debugMode") iss >> debugMode;
		else if (name == "sequenceBasePath") iss >> sequenceBasePath;
		else if (name == "sequenceName") iss >> sequenceName;
		else if (name == "resultsPath") iss >> resultsPath;
        else if (name == "precisionPath") iss >> precPath;
		else if (name == "frameWidth") iss >> frameWidth;
		else if (name == "frameHeight") iss >> frameHeight;
		else if (name == "seed") iss >> seed;
		else if (name == "searchRadius") iss >> searchRadius;
		else if (name == "svmC") iss >> svmC;
		else if (name == "svmBudgetSize") iss >> svmBudgetSize;
        else if (name == "multiSeq")
        {
            string sn;
            iss >> sn;
            mSeqNames.push_back(sn);
        }
		else if (name == "feature")
		{
			string featureName, kernelName;
			double param;
			iss >> featureName >> kernelName >> param;
			
			FeatureKernelPair fkp;
			
			if      (featureName == FeatureName(kFeatureTypeHaar)) fkp.feature = kFeatureTypeHaar;
			else if (featureName == FeatureName(kFeatureTypeRaw)) fkp.feature = kFeatureTypeRaw;
			else if (featureName == FeatureName(kFeatureTypeHistogram)) fkp.feature = kFeatureTypeHistogram;
            else if (featureName == FeatureName(kFeatureTypeHsv)) fkp.feature = kFeatureTypeHsv;
            else if (featureName == FeatureName(kFeatureTypeCircleHsv)) fkp.feature = kFeatureTypeCircleHsv;
            else if (featureName == FeatureName(kFeatureTypeCircleGrad)) fkp.feature = kFeatureTypeCircleGrad;
            else if (featureName == FeatureName(kFeatureTypeCircleRgb)) fkp.feature = kFeatureTypeCircleRgb;
            else if (featureName == FeatureName(kFeatureTypeCirclePatch)) fkp.feature = kFeatureTypeCirclePatch;
            else
			{
				cout << "error: unrecognised feature: " << featureName << endl;
				continue;
			}

            mfeatureName = featureName;
			
			if      (kernelName == KernelName(kKernelTypeLinear)) fkp.kernel = kKernelTypeLinear;
			else if (kernelName == KernelName(kKernelTypeIntersection)) fkp.kernel = kKernelTypeIntersection;
			else if (kernelName == KernelName(kKernelTypeChi2)) fkp.kernel = kKernelTypeChi2;
			else if (kernelName == KernelName(kKernelTypeGaussian))
			{
				if (iss.fail())
				{
					cout << "error: gaussian kernel requires a parameter (sigma)" << endl;
					continue;
				}
				fkp.kernel = kKernelTypeGaussian;
				fkp.params.push_back(param);
			}
			else
			{
				cout << "error: unrecognised kernel: " << kernelName << endl;
				continue;
			}
			
			features.push_back(fkp);
		}
	}
}

void Config::SetDefaults()
{

	quietMode = false;
	debugMode = false;
	
	sequenceBasePath = "";
	sequenceName = "";
	resultsPath = "";
    precPath = "";
	
	frameWidth = 320;
	frameHeight = 240;
	
	seed = 0;
	searchRadius = 30;
	svmC = 1.0;
	svmBudgetSize = 0;
	
	features.clear();
}

std::string Config::FeatureName(FeatureType f)
{
	switch (f)
	{
	case kFeatureTypeRaw:
		return "raw";
	case kFeatureTypeHaar:
		return "haar";
	case kFeatureTypeHistogram:
		return "histogram";
    case kFeatureTypeHsv:
        return "hsv";
    case kFeatureTypeCircleHsv:
        return "circle_hsv";
    case kFeatureTypeCircleGrad:
        return "circle_grad";
    case kFeatureTypeCircleRgb:
        return "circle_rgb";
    case kFeatureTypeCirclePatch:
        return "circle_patch";
	default:
		return "";
	}
}

std::string Config::KernelName(KernelType k)
{
	switch (k)
	{
	case kKernelTypeLinear:
		return "linear";
	case kKernelTypeGaussian:
		return "gaussian";
	case kKernelTypeIntersection:
		return "intersection";
	case kKernelTypeChi2:
		return "chi2";
	default:
		return "";
	}
}

ostream& operator<< (ostream& out, const Config& conf)
{
	out << "config:" << endl;
	out << "  quietMode          = " << conf.quietMode << endl;
	out << "  debugMode          = " << conf.debugMode << endl;
	out << "  sequenceBasePath   = " << conf.sequenceBasePath << endl;
	out << "  sequenceName       = " << conf.sequenceName << endl;
	out << "  resultsPath        = " << conf.resultsPath << endl;
    out << "  precisionPath      = " << conf.precPath << endl;
	out << "  frameWidth         = " << conf.frameWidth << endl;
	out << "  frameHeight        = " << conf.frameHeight << endl;
	out << "  seed               = " << conf.seed << endl;
	out << "  searchRadius       = " << conf.searchRadius << endl;
	out << "  svmC               = " << conf.svmC << endl;
	out << "  svmBudgetSize      = " << conf.svmBudgetSize << endl;
	
	for (int i = 0; i < (int)conf.features.size(); ++i)
	{
		out << "  feature " << i << endl;
		out << "    feature: " << Config::FeatureName(conf.features[i].feature) << endl;
		out << "    kernel:  " << Config::KernelName(conf.features[i].kernel) <<endl;
		if (conf.features[i].params.size() > 0)
		{
			out << "    params: ";
			for (int j = 0; j < (int)conf.features[i].params.size(); ++j)
			{
				out << " " << conf.features[i].params[j];
			}
			out << endl;
		}
	}
	
	return out;
}
